### PR TITLE
feat: support distinguishing escape sequences inside strings, fix multiline strings

### PR DESCRIPTION
**Pull request by amaanq**
Checklist:
- [x] All tests pass in CI.
- [x] There are sufficient tests for the new fix/feature. (Existing literals.txt reflects updates)
- [x] Grammar rules have not been renamed unless absolutely necessary.
- [x] The conflicts section hasn't grown too much.
- [x] The parser size hasn't grown too much (check the value of STATE_COUNT in src/parser.c). (1282 -> 1297)

### CODE REVIEW

**Comment by amaanq** [comment]
not sure how to make the CI run, but on my end running tree-sitter test and ./script/parse-examples is ok

**Comment by aryx** [comment]
cc @bmahe

**Comment by amaanq** [comment]
Note that I derived a lot of inspiration from the way the javascript parser handles escape sequences [here](https://github.com/tree-sitter/tree-sitter-javascript/blob/15e85e80b851983fab6b12dce5a535f5a0df0f9c/grammar.js#L906)

**Comment by bmahe** [comment]
Interesting changes!
@amaanq can we add some test cases to exercise the new cases being handled?

File: README.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
 tree-sitter-java
 ================
 
-[![Build/test](https://github.com/tree-sitter/tree-sitter-java/actions/workflows/ci.yml/badge.svg)](https://github.com/tree-sitter/tree-sitter-java/actions/workflows/ci.yml)
+[![CI](https://github.com/tree-sitter/tree-sitter-java/actions/workflows/ci.yml/badge.svg)](https://github.com/tree-sitter/tree-sitter-java/actions/workflows/ci.yml)
+[![Discord](https://img.shields.io/discord/1063097320771698699?logo=discord)](https://discord.gg/w7nTvsVJhm)
```
Comment (bmahe) [inline README.md:5]: This PR add advertisement for this discord server centered on tree-sitter.
I don't mind either way, but I want to make sure folks are fine about it

Comment (amaanq) [reply README.md:5]: It is pinned on the main tree-sitter repo's discussions

**Comment by amaanq** [comment]
> Interesting changes!
> @amaanq can we add some test cases to exercise the new cases being handled?

I do believe the multiline string changes are covered with the tests you added, but I can add some for the escape sequences soon :)

**Comment by amaanq** [comment]
@bmahe  Sorry, got caught up with life...added tests with the escape sequences

Let me know what you think of that.

**Comment by amaanq** [comment]
oops, didnt mean to remove the request

**Comment by amaanq** [comment]
Don't mean to be hasty or annoying...but I'd like to get this merged :slightly_smiling_face: 

ping @bmahe @aryx

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+module.exports = {
```
Comment (aryx) [inline .eslintrc.js:1]: Why did you add this in the repo? I don't think we do that for the other tree-sitter-xxx repos.

File: grammar.js [hunk 3]
```diff
@@ -1,5 +1,11 @@
-const DIGITS = token(sep1(/[0-9]+/, /_+/))
-const HEX_DIGITS = token(sep1(/[A-Fa-f0-9]+/, '_'))
+// deno-lint-ignore-file ban-ts-comment
```
Comment (aryx) [inline grammar.js:1]: It would have been better to do this estethic change in another PR and keep this PR
just for what's the title says.

**Review by aryx — CHANGES_REQUESTED** [review]

**Comment by bmahe** [comment]
@amaanq can we add tests that demonstrate the intent of this PR? 
The escape sequence provided in your latest commit is successfully parsed.

So what problem is this pr trying to resolve?


```bash
[bruno@buo tree-sitter-java]$ cat test.java 
String a = "\n\a\b\fhi im a piece of text\t\v and im some more text \\\'\"\?";
```
```bash
[bruno@buo tree-sitter-java]$ tree-sitter parse test.java 
Warning: You have not configured any parser directories!
Please run `tree-sitter init-config` and edit the resulting
configuration file to indicate where we should look for
language grammars.

(program [0, 0] - [2, 0]
  (local_variable_declaration [0, 0] - [0, 78]
    type: (type_identifier [0, 0] - [0, 6])
    declarator: (variable_declarator [0, 7] - [0, 77]
      name: (identifier [0, 7] - [0, 8])
      value: (string_literal [0, 11] - [0, 77]))))
```

File: .eslintrc.js (hunk 2 above)
Comment (amaanq) [reply .eslintrc.js:1]: it's a nice-to-have, and (with an additional ci/cd check) can enforce/check code styling/indentation/consistency, which, again, would be nice to have. I can remove this at your request

File: grammar.js (hunk 3 above)
Comment (amaanq) [reply grammar.js:1]: ok, sure

**Comment by amaanq** [comment]
> @amaanq can we add tests that demonstrate the intent of this PR? The escape sequence provided in your latest commit is successfully parsed.
> 
> So what problem is this pr trying to solve?
> 
> On the latest in the main branch:
> 
> ```shell
> [bruno@buo tree-sitter-java]$ cat test.java 
> String a = "\n\a\b\fhi im a piece of text\t\v and im some more text \\\'\"\?";
> ```
> 
> ```shell
> [bruno@buo tree-sitter-java]$ tree-sitter parse test.java 
> Warning: You have not configured any parser directories!
> Please run `tree-sitter init-config` and edit the resulting
> configuration file to indicate where we should look for
> language grammars.
> 
> (program [0, 0] - [2, 0]
>   (local_variable_declaration [0, 0] - [0, 78]
>     type: (type_identifier [0, 0] - [0, 6])
>     declarator: (variable_declarator [0, 7] - [0, 77]
>       name: (identifier [0, 7] - [0, 8])
>       value: (string_literal [0, 11] - [0, 77]))))
> ```

output on my end (with my changes):
ts p test.java

(program [0, 0] - [1, 0]
  (local_variable_declaration [0, 0] - [0, 78]
    type: (type_identifier [0, 0] - [0, 6])
    declarator: (variable_declarator [0, 7] - [0, 77]
      name: (identifier [0, 7] - [0, 8])
      value: (string_literal [0, 11] - [0, 77]
        (escape_sequence [0, 12] - [0, 14])
        (escape_sequence [0, 14] - [0, 16])
        (escape_sequence [0, 16] - [0, 18])
        (escape_sequence [0, 18] - [0, 20])
        (string_fragment [0, 20] - [0, 41])
        (escape_sequence [0, 41] - [0, 43])
        (escape_sequence [0, 43] - [0, 45])
        (string_fragment [0, 45] - [0, 68])
        (escape_sequence [0, 68] - [0, 70])
        (escape_sequence [0, 70] - [0, 72])
        (escape_sequence [0, 72] - [0, 74])
        (escape_sequence [0, 74] - [0, 76])))))

with this, editors can differentiate pieces of regular text vs escape sequences, like the image below:
image

the current parser doesn't allow for this. tree-sitter-python has a similar rule [here](https://github.com/tree-sitter/tree-sitter-python/blob/9e53981ec31b789ee26162ea335de71f02186003/grammar.js#L936), and tree-sitter-c [here](https://github.com/tree-sitter/tree-sitter-c/blob/7175a6dd5fc1cee660dce6fe23f6043d75af424a/grammar.js#L959)

File: .eslintrc.js (hunk 2 above)
Comment (aryx) [reply .eslintrc.js:1]: I'm ok adding this, it's just that it should be in a different PR. We try to have a PR doing one thing and one thing only.

**Review by aryx — CHANGES_REQUESTED** [review]
Please make a separate PR for the styling/spacing/ and all the other stuff
that are independent of what the title of this PR says.

In this PR we should just see modifications to distinguish escape sequence inside strings, and related test files, and nothing else.
One this is merged you can make another PR to add those eslint and ci checks for styling issue.

**Comment by amaanq** [comment]
> Please make a separate PR for the styling/spacing/ and all the other stuff that are independent of what the title of this PR says.
> 
> In this PR we should just see modifications to distinguish escape sequence inside strings, and related test files, and nothing else. One this is merged you can make another PR to add those eslint and ci checks for styling issue.

Gotcha, I believe I've removed everything unrelated to the point of the PR. Anything else @aryx?

**Comment by aryx** [comment]
There's a tons of stuff not related to the PR title still. all those semicolons, $ to _, etc.
Those are good changes, but they need to be in a separate PR.
The modifications in this PR should just be related to the title of the PR, so only the stuff that help distinguish escape sequences
inside strings

File: .gitattributes [hunk 4]
```diff
@@ -1,2 +1,6 @@
 /src/** linguist-vendored
 /examples/* linguist-vendored
+
+src/grammar.json -diff
```
Comment (aryx) [inline .gitattributes:4]: unrelated

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-name: Build/test
+name: CI
```
Comment (aryx) [inline .github/workflows/ci.yml:1]: unrelated

File: .gitignore [hunk 6]
```diff
@@ -1,18 +1,8 @@
...
-/examples/RxJava
-/target/
-*.a
-*.dylib
-*.so
-*.o
 bindings/c/*.h
 bindings/c/tree-sitter-*.pc
-.build/
+/build
```
Comment (aryx) [inline .gitignore:5]: unrelated

File: grammar.js [hunk 7]
```diff
@@ -1,5 +1,5 @@
-const DIGITS = token(sep1(/[0-9]+/, /_+/))
-const HEX_DIGITS = token(sep1(/[A-Fa-f0-9]+/, '_'))
+const DIGITS = token(sep1(/[0-9]+/, /_+/));
```
Comment (aryx) [inline grammar.js:1]: unrelated.
All those semicolons, space change, comma, ... all of that is unrelated.

File: grammar.js [hunk 8]
```diff
@@ -91,83 +91,112 @@ module.exports = grammar({
...
+    string_literal: $ => choice($._string_literal, $._multiline_string_literal),
+    _string_literal: $ => seq(
+      '"',
+      repeat(choice(
+        $.string_fragment,
+        $.escape_sequence,
+      )),
+      '"',
     ),
+    _multiline_string_literal: $ => seq(
```
Comment (aryx) [inline grammar.js:164]: I just want to see those modifications in the PR.

**Review by aryx — CHANGES_REQUESTED** [review]

File: grammar.js (hunk 7 above)
Comment (amaanq) [reply grammar.js]: removed now

File: grammar.js (hunk 8 above)
Comment (amaanq) [reply grammar.js:165]: I agree, my editor has an auto format on save which added the semicolons/commas and such, it is also a bit irking having the inconsistency of some comma sep items end in commas and some not. After this I'll submit a general PR adding the linting and other fixes across the board if whatever was in there before is ok w/ you.

**Comment by amaanq** [comment]
tidied up history a bit

**Comment by amaanq** [comment]
ping @aryx

File: grammar.js [hunk 9]
```diff
@@ -144,28 +144,58 @@ module.exports = grammar({
...
+    //
+    string_fragment: $ =>
+      token.immediate(prec(1, /[^"\\]+/)),
+    _multiline_string_fragment: () =>
+      prec.right(choice(
+        /[^"]+/,
+        seq(/"[^"]*/, repeat(/[^"]+/))
+      )),
+
+    _escape_sequence: $ =>
```
Comment (aryx) [inline grammar.js:185]: would have been good to add in comment a link to the javascript grammar that you took as inspiration
for those changes.

**Review by aryx — APPROVED** [review]
Thx, looks good now.

File: grammar.js (hunk 9 above)
Comment (amaanq) [reply grammar.js:185]: Ahh, fair, but I use this with almost every grammar I've made, would be a bit cumbersome :sweat_smile: I like to think of it as a more universal solution for tree-sitter parsers needing this feature/rule

File: grammar.js [hunk 10]
```diff
@@ -144,28 +144,58 @@ module.exports = grammar({
...
+      )),
+      '"""'
+    ),
+    // Workaround to https://github.com/tree-sitter/tree-sitter/issues/1156
+    // We give names to the token() constructs containing a regexp
+    // so as to obtain a node in the CST.
+    //
+    string_fragment: $ =>
+      token.immediate(prec(1, /[^"\\]+/)),
+    _multiline_string_fragment: () =>
```
Comment (ahumenberger) [inline grammar.js:179]: @amaanq Is there a specific reason why `_multiline_string_fragment` is not of the form `token(...)`? We had ` text_block: $ => token(...)` before.

The consequence is that the `_multiline_string_fragement` node is now a product type instead of a leaf type.

Compare 
```
  {
    "type": "multiline_string_fragment",
    "named": true,
    "fields": {}
  }
```
```
 {
   "type": "text_block",
   "named": true
 }
```

I'm writing an "AST generator" generator which generates the CST into an AST based on those node types, and I'm acting differently based on which type of node it is, that is, sum type, product type or leaf type. And the behavior of the grammar changed compared to the previous version.

Comment (ahumenberger) [reply grammar.js:179]: Just tried adding the `token` before, but then a test fails. With `token.immediate` it works though. I'm not familiar enough yet with the grammar rules if that would be okay.

Comment (amaanq) [reply grammar.js:179]: I'm not gonna lie, I forgot why. I use this style in several grammars of mine and I derived inspiration from another grammar.. `token.immediate` can have weird side effects causing annoying edge case bugs so be careful with that one.

Comment (amaanq) [reply grammar.js:179]: Try token(prec(-1 or prec(1 and see how those vary
